Optimizing Stacked Borrows (part 1?): Cache locations of Tags in a Borrow Stack

saethlin opened the pull request:
Before this PR, a profile of Miri under almost any workload points quite squarely at these regions of code as being incredibly hot (each being ~40% of cycles):

https://github.com/rust-lang/miri/blob/dadcbebfbd017aac2358cf652a4bd71a91694edc/src/stacked_borrows.rs#L259-L269

https://github.com/rust-lang/miri/blob/dadcbebfbd017aac2358cf652a4bd71a91694edc/src/stacked_borrows.rs#L362-L369

This code is one of at least three reasons that stacked borrows analysis is super-linear: These are both linear in the number of borrows in the stack and they are positioned along the most commonly-taken paths.

I'm addressing the first loop (which is in `Stack::find_granting`) by adding a very very simple sort of LRU cache implemented on a `VecDeque`, which maps recently-looked-up tags to their position in the stack. For `Untagged` access we fall back to the same sort of linear search. But as far as I can tell there are never enough `Untagged` items to be significant.

I'm addressing the second loop by keeping track of the region of stack where there could be items granting `Permission::Unique`. This optimization is incredibly effective because `Read` access tends to dominate and many trips through this code path now skip the loop entirely.

These optimizations result in pretty enormous improvements:
Without raw pointer tagging, `mse` 34.5s -> 2.4s, `serde1` 5.6s -> 3.6s
With raw pointer tagging, `mse` 35.3s -> 2.4s, `serde1` 5.7s -> 3.6s

And there is hardly any impact on memory usage:
Memory usage on `mse` 844 MB -> 848 MB, `serde1` 184 MB -> 184 MB (jitter on these is a few MB).

camelid commented on src/stacked_borrows.rs:
```diff
@@ -71,15 +71,35 @@ impl fmt::Debug for Item {
+    /// position in the stack.
+    cache: FxHashMap<PtrId, usize>,
+    /// `Untagged` may occur multiple times so we store it outside of the map.
+    untagged: Vec<usize>,
```
Perhaps this is pre-existing, but it might be a good idea to use a newtype here instead of `usize`.

saethlin replied:
How would that help? These values exist only to be used as indexes into `borrows`.

camelid replied:
Ah, then it'd be good to add a comment about that. Reading this field definition, I had a couple of confusion points:

- What is the meaning of the `usize`?
- Is this ordered or could it be a set instead?

saethlin replied:
Your newtype suggestion is looking better and better as I try things out :sweat_smile:

RalfJung commented:
Thanks a lot for this PR! Those are some impressive wins given the conceptually simple changes. :)

> I'm addressing the second loop by keeping track of the region of stack where there could be items granting Permission::Unique. This optimization is incredibly effective because Read access tends to dominate and many trips through this code path now skip the loop entirely.

Presumably this also has basically no memory cost, so I wonder how much just doing this would help?

> But as far as I can tell there are never enough Untagged items to be significant.

So could we save some memory by removing this part?

> These optimizations result in a 30% improvement in the the serde1 benchmark and an 85% (~6x) improvement in the mse benchmark. Which is pretty great because I didn't even develop these changes by running against those programs.

Note that I have no idea how suited these benchmarks are. It might be good to do at least superficial benchmarking of the code in `tests/run-pass/stacked-borrows` to avoid surprises for more "normal" code.

> But as far as I can tell from some very casual instrumenting, the root cause of the stacked borrows memory usage is that borrow stacks grow over time without bound and a RangeMap<Stack> will keep duplicating Stacks. The Clone+PartialEq activity that this generates eventually dominates as stacks grow, because that's O(n) where the HashMap optimization here gets darn close to O(1) for all other operations. I think it may be possible to ease this by nesting stacks, something like parent: Option<Rc<Stack>>, to avoid the expensive Clone operation until we actually need to modify the older elements.

That was one of my hypotheses for the main culprit. The other one is that even if RangeMap is fitting to the access pattern (so we don't keep splitting and merging subranges), just the fact that the stacks keep growing will be expensive in the long run -- even if we have perfect sharing, that's still quadratic since we keep iterating a larger and larger vector.

saethlin commented:
> Thanks a lot for this PR! Those are some impressive wins given the conceptually simple changes. :)

Hah! I was afraid these would look prohibitively complex.

> So could we save some memory by removing this part?

We could, if we always tag raw pointers. Without `-Zmiri-tag-raw-pointers`, this data structure returns indexes in `find_granting` and we need the right index to invalidate the right part of the stack. Or at least that's what I've been able to glean so far.

As a sidenote: This is definitely the reason `-Zmiri-tag-raw-pointers` is faster with this PR. Without it, there are still linear searches of the untagged stack in `find_granting`, and the number of Untagged borrows is up to about half that of the Tagged ones in large stacks.

> Note that I have no idea how suited these benchmarks are.

Yeah. I'll cook up a little script to run through some more reasonable benchmarking code with the various SB options.

> just the fact that the stacks keep growing will be expensive in the long run -- even if we have perfect sharing, that's still quadratic since we keep iterating a larger and larger vector.

I suspect all iteration can be efficiently amortized out. If the parent stack from a split is stored in an `Rc`/`Arc` we can use `ptr_eq` to do a constant-time parent comparison. But there also other questions about that implementation turns out on a real workload (I haven't even started on it yet).

RalfJung commented:
> We could, if we always tag raw pointers. Without -Zmiri-tag-raw-pointers, this data structure returns indexes in find_granting and we need the right index to invalidate the right part of the stack. Or at least that's what I've been able to glean so far.

Well, we could always use a linear scan instead, right? That will only slow down raw ptr accesses, and save some memory. This might or might not be a reasonable trade-off, not sure.

> Without it, there are still linear searches of the untagged stack in find_granting, and the number of Untagged borrows is up to about half that of the Tagged ones in large stacks.

That is odd, since the search should stop at the first untagged item it finds -- except when it is looking for a writeable item, then it will skip all the read-only untagged items.

saethlin commented:
> That is odd, since the search should stop at the first untagged item it finds

You're right. I don't know what I was looking at before, but this is a good insight. Nearly every access of an Untagged item wants the topmost Untagged item (easily demonstrated with some `println!` and `cargo miri run | counts`. Removing the Vec of untagged indices has a ~3% improvement in RSS on `mse`, no change on `serde1`. But I see a ~50% improvement in the runtime of `mse` and a similarly-large regression on `serde1`. Which is itself interesting; some workloads are very Untagged-heavy but don't have an Untagged near the top of the stack.

So I of course slapped a trivial cache on it, and now it's an improvement across the board.

I _was_ planning on using `core` as my benchmarks suite, but of course the `join` test is failing :cry: so I'm going to collect some interesting cases from third-party crates I've run `cargo miri test` on.

RalfJung commented:
> I was planning on using core as my benchmarks suite, but of course the join test is failing

https://github.com/rust-lang/rust/pull/92121 should help with that.

bors commented:
:umbrella: The latest upstream changes (presumably #1945) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented:
I'm still slowly working on this.

I tried running miri-test-libstd over `core` with `-Zmiri-tag-raw-pointers` and I killed the process after 2 hours (using 20 GB). It spent over an hour doing _something_ after printing
```
Failures:
```
So I have absolutely no idea what was going on there. It would be cool to have a way to peer inside miri and see what code it is executing, best guess is something in the test harness that ends up quadratic or worse inside miri due to the nature of stacked borrows.

I also tried running it the tests for `core` with this branch, and the memory usage started growing rapidly at the same stage in the run, I killed the process when it hit 100 GB. Neither of these runs completed, so I don't think there is a 5x memory blowup from this PR, it seems likely that memory would have only grown to this level if I'd let the current tip of master run long enough.

At the moment I'm chewing on ideas to do something about this. I don't think a patch that makes running the standard library tests with tag-raw-pointers a non-starter is a good idea.

RalfJung commented:
> t would be cool to have a way to peer inside miri and see what code it is executing,

Yeah that would be cool... Cc https://github.com/rust-lang/miri/issues/1782.
The issue you describe sounds familiar as well: https://github.com/rust-lang/miri/issues/1780. However it is odd that one of the tests would fail...

RalfJung commented:
> I'm still slowly working on this.

Marking as "waiting for author" based on above comment. Please let me know if that is inaccurate. :)

saethlin commented:
That is accurate.

I'm currently focusing on other things, but I'm becoming increasingly convinced that the only good way out of this borrow stack growth situation is to GC the borrow stacks. But even if that works, it would require that Miri be notified when a pointer or reference goes away, which of course bumps up against ptr-int-ptr casts. So overall I feel like I don't have any ideas which are easy enough to implement that they rise up my priority list, and the hard ideas I have are stepping right into the most unstable parts of Miri.

bors commented:
:umbrella: The latest upstream changes (presumably #2030) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented:
I've been thinking about this problem for a while, and I finally decided to code up my idea for a super simple limited-size cache. I've rebased in the new implementation and updated the top comment.

This is freaking awesome, can't believe I didn't try this approach before. I think there are a lot of clear ways that this could be micro-optimized, but what I'm seeing in profiles makes that work hard to justify.

saethlin commented:
@RalfJung @oli-obk this is not waiting-on-author anymore, it's looking-for-advice-on-factoring :)

saethlin commented:
Examples where this doesn't help as much

regex: the stack merging code seems to be really active here. ![regex==1.5.5](https://user-images.githubusercontent.com/12105168/168450814-b5dca1c3-85c9-42ab-a519-745f3dcf699b.svg) 

unicode-xid: the the cache hit rate in `find_granting` is still really high, but we keep having to do searches of tends of thousands of tags. I bumped the cache size to 32 with no impact, unclear if making it super big would help.
![unicode-xid==0.2.3](https://user-images.githubusercontent.com/12105168/168450854-c438c857-0683-4d36-b9ae-36aa75616f58.svg)

RalfJung commented:
> My biggest concern at this point is code organization.

Yeah, I agree. As a starter, actually documenting the data structure invariants and how exactly the cache relates to the "live" data would be good. :) Right now, to review this I would have to reverse engineer all that. In fact, I think we should have an actual Rust function that checks this invariant, i.e., that checks whether the cache is consistent with the "live" data, and we should have a compile-time flag to enable run-time checking that the cache does not get out-of-sync with the live data. Such a machine-readable invariant would probably also help clarify/complement the human-readable comments.

As for factoring things, the first "obvious" proposal would be to move `Stack` to its own file and make sure the public API is such that caching can never be broken. Is that possible?

saethlin commented:
> we should have a compile-time flag 

:facepalm: missed the compile-time part of this and I implemented a `-Z` flag the whole time thinking "this makes much more sense as a compile time setting".

What makes sense as a way to implement this? a `--cfg`? A build-time environment variable check with `env!`? I feel like we would want this turned on when running tests, so normally I would reach for `cfg(debug_assertions)` but that's turned on by default for `./miri install` so I think that's the wrong choice?

RalfJung commented:
Hm, not sure if we truly want it for our tests, given that it'll probably make things hugely slow.

Not sure what the best way is to do this. A `--cfg` cannot easily be set from cargo, can it? It seems too expensive for `debug_assertions`. So `env!` sounds like the least bad option?

RalfJung commented:
Or maybe a cargo feature flag?

bors commented:
:umbrella: The latest upstream changes (presumably #2139) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented:
I just rebased up, collapsed down into two commits because I think it tells a cleaner story, and edited out another optimization I was working on that isn't part of the caching approach that this PR is about.

oli-obk commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+        Ok(())
+    }
+
+    pub fn drain_rev(&mut self, range: std::ops::Range<usize>) -> DrainRev<'_> {
+        DrainRev { start: range.start, stack: self, cursor: range.end - 1 }
+    }
```
This is only ever used with a range to `self.len()`, maybe just pass a start and compute the end here? the `- 1` offset made me wonder if it could bite us somewhere.

oli-obk commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cursor >= self.start {
+            self.cursor -= 1;
+            Some(self.stack.borrows[self.cursor + 1].clone())
```
```suggestion
            let elem = self.stack.borrows[self.cursor].clone();
            self.cursor -= 1;
            Some(elem)
```

oli-obk reviewed (changes requested):
Just some style nits. The magic off by one or twos are not ideal, but you documented them, so should be fine.

RalfJung commented on ci.sh:
```diff
@@ -20,11 +20,11 @@ function run_tests {
   fi
 
-  ./miri test --locked
+  ./miri test --locked --features=expensive-debug-assertions
```
This will rebuild miri since the original `./miri build` was with a different feature set.

Maybe put these flags into a variable and pass them to all invocations?
In fact, we already have `CARGO_EXTRA_FLAGS="--all-features"`, isn't that enough?

saethlin replied:
Yes I think so. I just didn't notice that in the CI script.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -321,15 +298,15 @@ impl<'tcx> Stack {
         global: &GlobalStateInner,
         alloc_history: &mut AllocHistory,
-    ) -> InterpResult<'tcx> {
+    ) -> InterpResult<'static> {
```
Recently in rustc all `InterpResult<'static>` were changed to `'tcx`, so please also suck to `'tcx` here.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -321,15 +298,15 @@ impl<'tcx> Stack {
     /// currently checking.
     fn check_protector(
-        item: &Item,
+        item: Item,
```
Any specific reason for this change?

saethlin replied:
Encapsulation. If `Stack`'s interface returns `&Item` anywhere, it is leaking that it actually contains `Item`s and that means that it needs to use array-of-structs instead of struct-of-arrays.

RalfJung replied:
This is an argument type though. Anyone who has an `Item` can just pass `&item` to satisfy the old type. So I don't see how this helps with encapsulation.

saethlin replied:
:facepalm: you're right, there's no point in propagating the type change like this.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+                    .find_map(|(idx, item)| if item.perm.grants(access) { Some(idx) } else { None }),
+            SbTag::Tagged(_) => {
+                // Search the cache for the tag we're looking up
+                if let Some(cache_idx) = self.cache.tags.iter().position(|t| *t == tag) {
```
So we do a linear search of the cache here?
I am quite surprised that that is beneficial...^^

oli-obk replied:
If the cache is small, this is a common cache optimization

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+use crate::stacked_borrows::{AccessKind, Item, Permission, SbTag};
+use std::ops::Range;
+
+const CACHE_LEN: usize = 32;
```
Please add a comment giving at least some inkling of how this number was chosen. I assume there's a tradeoff here, bigger cache is slower to look up in?

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+        self.borrows.len()
+    }
+
+    pub fn disable_uniques_above<V: FnMut(Item) -> crate::InterpResult<'static>>(
```
This function should probably have a doc comment.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -82,16 +83,34 @@ impl fmt::Debug for Item {
     /// * Above a `SharedReadOnly` there can only be more `SharedReadOnly`.
     /// * Except for `Untagged`, no tag occurs in the stack more than once.
     borrows: Vec<Item>,
+    cache: VecDeque<(SbTag, usize)>,
```
Sorry for this, I must have written this a while ago and GH decided to now add it to my new review...

bors commented:
:umbrella: The latest upstream changes (presumably #2032) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #2161) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,258 @@
+        Ok(())
+    }
+
+    pub fn drain_rev(&mut self, range: std::ops::Range<usize>) -> DrainRev<'_> {
```
And feedback on if this is even the right way to abstract things. It feels all sorts of leaky to me.

RalfJung replied:
I have a hard time even understanding what the contract is, so it's hard to comment on whether it is the right one.^^ That's why I asked for doc comments.

bors commented:
:umbrella: The latest upstream changes (presumably #2166) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -388,10 +365,9 @@ impl<'tcx> Stack {
             let first_incompatible_idx = self.find_first_write_incompatible(granting_idx);
-            for item in self.borrows.drain(first_incompatible_idx..).rev() {
-                trace!("access: popping item {:?}", item);
+            for item in self.drain_rev(first_incompatible_idx..self.len()) {
```
This is the only call site for `drain_rev`, right?

I used `drain` because it seemed elegant, but AFAIK this is equivalent to
```rust
while self.borrows.len() > first_incompatible_idx {
  let item = self.borrows.pop().unwrap();
  // ...
}
```
so maybe that works better for the stack-with-a-cache?
```

saethlin replied:
I think this API is less obtuse and overall it's a win. It's still hampered a bit by the fact that we need to put the cache back into a valid state once we're done with this operation, but perhaps I've named/documented things a bit better now so it's clearer?

RalfJung commented:
I think what would be useful for future development of Stacked Borrows is to be able to turn off the cache, so that one can just work on the changed semantics (e.g. wildcard pointers) without worrying about the cache -- and then add back the cache. Any good ideas for what would be the best way to make that possible?

RalfJung commented:
One option might be to put the cache behind a cargo feature and add that feature to default-features. Then when working on Stacked Borrows one could temporarily remove the feature from default-features and not worry about caching at first.

saethlin commented:
The caching system that I want to use for `find_granting` is not particularly difficult to factor out. The logic is already broken into chunks that have obvious behaviors, in `find_granting` you call a cache-querying function that just returns `None` when the cache is off, and in `insert` you don't update it.

I'm not entirely sure what to do right now about the other linear scan of the borrow stack. Operations like disabling `Unique`s above an item and finding the first write-incompatible item do not have such obvious factorings (though this PR does not deal with `find_first_write_incompatible`, I now have a benchmark that stresses it so I'd like to prevent backing the implementation into a corner so that it can be addressed later).

I'm going to think about this for a bit today and get back to this.

RalfJung commented:
Ah, so those will not have fallback paths for cache misses, they will always use some kind of fast path?

bors commented:
:umbrella: The latest upstream changes (presumably #2196) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -406,8 +383,13 @@ impl<'tcx> Stack {
-                trace!("access: popping item {:?}", item);
+            let items_to_remove = self.len() - first_incompatible_idx;
+            // This mutation of the stack temporarily jumbles some internal state. This guard
+            // repairs that state on drop, and while it exists it only permits operations which
```
So, we assume the guard is never leaked?

I guess that's fine, though a closure-based API seems better to me.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -466,7 +444,8 @@ impl<'tcx> Stack {
         })?;
 
         // Step 2: Remove all items.  Also checks for protectors.
```
This doesn't "remove" any more.

I was thinking of this as being similar to the popping on access... but since the memory goes away I guess it is okay to implement it very differently.

saethlin replied:
What do you mean? I'm pretty sure the caching system honors removal properly, or are you referring to something with Wildcard?

RalfJung replied:
What I meant is they don't actually remove things from the stack, they just iterate over the entire stack and do something for each element.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,307 @@
+
+    #[cfg(feature = "stack-cache")]
+    fn find_granting_cache(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        if self.borrows.len() > CACHE_LEN {
```
You can avoid some rightwards drift here by negating the condition and doing an early `return None`. Also please add a comment explaining the purpose of this.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,307 @@
+        }
+
+        // The above insert changes the meaning of every index in the cache >= new_idx, so now
+        // we need to find every one of those indexes and increment it.
```
This makes me wonder if we should do a bit more work to do *less* work when the borrow stack is small -- smaller than the cache, even. It seems to me like this cache will massively slow down operations on stacks that are about as large as the cache, due to all the extra bookkeeping.

OTOH your benchmarks show a clear speedup, so I guess the lesson is that the cost is so dominated by the large stacks that making the small stacks slower is fine? Or that this is all O(1) and complexity theory is actually right? Or something like that?

saethlin replied:
First off, I agree that this loop is a performance hazard. I think I mentioned that at your request in the documentation of `CACHE_LEN`.

I am pretty sure we get away with this behavior because the cache is a crazy efficient data structure, especially by comparison to the rest of Miri. It's just loops over a small(ish) array of constant length and modern processors love code like that. It _is_ O(1), but I hesitate to use that as a justification for anything because not all programs push Stacked Borrows into complexity theory territory. For example, if a program somehow had a totally pathological borrowing pattern it could end up with mostly cache misses and then we would really see these O(1) cache operations surface. But to be clear, I have yet to find such a program. I checked in the most pathological programs I could find as other benchmarks, and _at worst_ these cache operations amount to 2.2% of Miri's runtime. Which is minuscule compare to what runtime the cache is saving.

This is so messy from a theoretical perspective, and we should really have better benchmarking (something like a fork of rustc-perf would be cool), but also this is such a dramatic win on so many programs that I would prefer to land this optimization, then do a part 2 with more unambiguous but smaller wins.

RalfJung replied:
That sounds totally fair, I just wanted to be sure. :) 

If you could summarize this in one sentence for a comment in the code, that would be great. Basically just acknowledging that yes this might seem like a waste for tiny stacks but in practice it is a huge win since (a) "modern processors love code like that" and (b) runtime is dominated by large stacks.

RalfJung commented:
Yes, I like that API. :)

I did screw you over with that permissive provenance PR though, causing a lot of conflicts. I consider that a very important feature (it lets us remove this  untagged nonsense!) so I put the burden of making it compatible with your cache on you... I hope that's okay.

saethlin commented:
> I hope that's okay.

It's quite okay :)

I'm just spending my free time on other things at the moment, so I may not get around to another iteration on this for a few days.

saethlin commented:
I'm getting denied by clippy:
> error: this returns a `Result<_, ()>`

But isn't that same signature currently on the master branch? I'm happy to add an allow here, but it feels odd to be adding the allow in this PR.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -51,6 +51,13 @@ struct StackCache {
 impl StackCache {
+    /// When a tag is used, we call this function to add or refresh it in the cache.
+    ///
+    /// We use position in the cache to represent how recently a tag was used; the first position
```
```suggestion
    /// We use the position in the cache to represent how recently a tag was used; the first position
```

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,363 @@
+    /// index is given it means the match was *not* in the known part of the stack.
+    /// `Ok(None)` indicates it matched the "unknown" part of the stack.
+    /// `Err` indicates it was not found.
+    pub fn find_granting(
```
Clippy complains because this is now `pub`. Maybe make it `pub(super)` or so?

RalfJung commented on ui_test/Cargo.toml:
```diff
@@ -14,3 +14,7 @@ crossbeam = "0.8.1"
 serde_json = "1.0"
+
+[features]
+# Doesn't do anything, but the miri script wants to pass the same flags to ui_test and miri itself
```
The flags it passes are `--all-features` though so why is this needed?

saethlin replied:
You are correct. I was trying to pass a specific feature before.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    pub fn set_unknown_bottom(&mut self, tag: SbTag) {
+        self.borrows.clear();
```
No cache to clear here?

saethlin replied:
The cache can't be cleared, because it doesn't have a dynamic length so I've tried to explain in a commit why not clearing it is okay.

I get that this is a bit of a complicated requirement to uphold, but on the other hand that is why we have `verify_cache_consistency`. Giving the cache a length would be similar to making it an `arrayvec` and I'm not sure that's a win.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+    /// - There are no Unique tags outside of first_unique..last_unique
+    #[cfg(feature = "expensive-debug-assertions")]
+    fn verify_cache_consistency(&self) {
+        if self.borrows.len() > CACHE_LEN {
```
```suggestion
        // Only a full cache needs to be valid. Also see the comments in find_granting_cache
        // and set_unknown_bottom.
        if self.borrows.len() > CACHE_LEN {
```
The one thing I am surprised by is that this is not `>=`. But it is consistent with `find_granting_cache` so seems fine.

saethlin replied:
Nah, I'll bump it up to `>=`. The `==` case ought to be valid.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -900,11 +826,14 @@ trait EvalContextPrivExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     .as_ref()
                     .expect("we should have Stacked Borrows data")
                     .borrow_mut();
+                let mut current_span = this.machine.current_span();
```
What changed that makes it so that we need a new `current_span` here, rather than reusing the old one as before?

saethlin replied:
:thinking: me being bad at rebasing probably

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -553,7 +480,8 @@ impl<'tcx> Stack {
         })?;
 
         // Step 2: Remove all items.  Also checks for protectors.
```
```suggestion
        // Step 2: Consider all items removed. This checks for protectors.
```

RalfJung commented on Cargo.toml:
```diff
@@ -50,3 +50,8 @@ rustc_private = true
+
+[features]
+default = ["stack-cache"]
+expensive-debug-assertions = []
```
```suggestion
# Will be enabled on CI via `--all-features`.
expensive-debug-assertions = []
```

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -23,6 +23,10 @@ use crate::*;
+pub mod stack;
+use stack::Stack;
+
+pub type PtrId = NonZeroU64;
```
```suggestion
```
This type no longer exists.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+    /// we never have the unknown-to-known boundary in an SRW group.
+    unknown_bottom: Option<SbTag>,
+
+    /// A small LRU cache of searches of the borrow stack. This only caches accesses of `Tagged`,
+    /// because those are unique in the stack.
```
```suggestion
    /// A small LRU cache of searches of the borrow stack.
```
Nothing else even exists in the stack any more.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+/// This maps tags to locations in the borrow stack. Any use of this still needs to do a
+/// probably-cold random access into the borrow stack to figure out what `Permission` an
+/// `SbTag` grants. We could avoid this by also storing the `Permission` in the cache, but
+/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
```
```suggestion
/// most lookups into the cache are immediately followed by access of the full borrow stack anyway.
///
/// It may seem like maintaining this cache is a waste for small stacks, but
/// (a) iterating over small fixed-size arrays is super fast, and (b) empirically this helps *a lot*,
/// probably because runtime is dominated by large stacks.
/// arrays is super fast
```

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+        }
+
+        // If we didn't find the tag in the cache, fall back to a linear search of the
+        // whole stack, and add the tag to the stack.
```
```suggestion
        // whole stack, and add the tag to the cache.
```
I presume?

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+                self.cache.tags.swap(0, 1);
+                self.cache.idx.swap(0, 1);
+            } else if cache_idx > 1 {
+                self.cache.add(stack_idx, tag);
```
This doesn't "move" it to the top, it adds it to the top and shifts everything, so now it is in the cache twice, right?

saethlin replied:
Oh goodness, you're right. I threw together the better implementation, and it is a bit surprising to me that this doesn't seem to matter. It's possible that because I've already special-cased lookups of the first 2 elements in the cache, which should be by far the most commonly used, this duplication just doesn't happen very much? So at the moment, I just explained in a comment what's going on here as best I can, and I favor keeping the implementation simple.

RalfJung replied:
Sounds good.

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -0,0 +1,372 @@
+        let unique_range = 0..self.len();
+
+        if disable_start <= unique_range.end {
+            // add 1 so we don't disable the granting item
```
I don't understand this comment.

saethlin replied:
Because it's leftover from when the argument to that function was the granting index. Easy fix.

RalfJung commented:
Other than my last rounds of comments, I think this is good to go. :)

RalfJung commented on src/stacked_borrows/stack.rs:
```diff
@@ -42,6 +41,11 @@ pub struct Stack {
+///
+/// It may seem like maintaining this cache is a waste for small stacks, but
+/// (a) iterating over small fixed-size arrays is super fast, and (b) empirically this helps *a lot*,
+/// probably because runtime is dominated by large stacks.
+/// arrays is super fast
```
```suggestion
/// probably because runtime is dominated by large stacks.
```
Did GH add that back? It appeared in my suggestion somehow but I edited it out... very strange.

saethlin replied:
Yeah, it was part of the suggestion :thinking:

RalfJung commented:
Stacked Borrows go brrrrrr
@bors r+

bors commented:
:pushpin: Commit b004a03bdb6affd7b07322114967bd1693208a69 has been approved by `RalfJung`

<!-- @bors r=RalfJung b004a03bdb6affd7b07322114967bd1693208a69 -->
<!-- homu: {"type":"Approved","sha":"b004a03bdb6affd7b07322114967bd1693208a69","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit b004a03bdb6affd7b07322114967bd1693208a69 with merge cfad9d12f32634160e5aaecc41fb4097658b2bb2...
<!-- homu: {"type":"BuildStarted","head_sha":"b004a03bdb6affd7b07322114967bd1693208a69","merge_sha":"cfad9d12f32634160e5aaecc41fb4097658b2bb2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/7169604894?check_suite_focus=true)
Approved by: RalfJung
Pushing cfad9d12f32634160e5aaecc41fb4097658b2bb2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/7169604894?check_suite_focus=true"},"merge_sha":"cfad9d12f32634160e5aaecc41fb4097658b2bb2"} -->
